README.MD
# 🔐 JSON Server Auth

JWT authentication middleware for **[JSON Server](https://github.com/typicode/json-server)**

Because you also need a fake **authentication & authorization flow** for your prototyping.

## Getting started

Install **both** JSON Server and JSON Server Auth :

```bash
# NPM
npm install -D json-server json-server-auth

# Yarn
yarn add -D json-server json-server-auth
```

Create a `db.json` file with a `users` collection :

```json
{
  "users": []
}
```

Start JSON server (with _JSON server Auth_ as middleware) :

```bash
json-server db.json -m ./node_modules/json-server-auth
# with json-server installed globally and json-server-auth installed locally
```

##### 📢 but wait !

As a convenience, **`json-server-auth`** CLI exposes `json-server` bundled with its middlewares :

```bash
json-server-auth db.json
# with json-server-auth installed globally
```

_It exposes and works the same for all [JSON Server flags](https://github.com/typicode/json-server#cli-usage)._

## Authentication flow 🔑

JSON Server Auth adds a simple [JWT based](https://jwt.io/) authentication flow.

### Register 👥

Any of the following routes registers a new user :

- **`POST /register`**
- **`POST /signup`**
- **`POST /users`**

**`email`** and **`password`** are required in the request body :

```http
POST /register
{
  "email": "olivier@mail.com",
  "password": "bestPassw0rd"
}
```

The password is encrypted by [bcryptjs](https://github.com/dcodeIO/bcrypt.js).

The response contains the JWT access token (expiration time of 1 hour), and the user data (without the password) :

```http
201 Created
{
  "accessToken": "xxx.xxx.xxx",
  "user": {
    "id": 1,
    "email": "olivier@mail.com"
  }
}
```

###### Other properties

Any other property can be added to the request body without being validated :

```http
POST /register
{
  "email": "olivier@mail.com",
  "password": "bestPassw0rd",
  "firstname": "Olivier",
  "lastname": "Monge",
  "age": 32
}
```

###### Update

Any update to an existing user (via `PATCH` or `PUT` methods) will go through the same process for `email` and `password`.

### Login 🛂

Any of the following routes logs an existing user in :

- **`POST /login`**
- **`POST /signin`**

**`email`** and **`password`** are required, of course :

```http
POST /login
{
  "email": "olivier@mail.com",
  "password": "bestPassw0rd"
}
```

The response contains the JWT access token (expiration time of 1 hour), and the user data (without the password) :

```http
200 OK
{
  "accessToken": "xxx.xxx.xxx",
  "user": {
    "id": 1,
    "email": "olivier@mail.com",
    "firstname": "Olivier",
    "lastname": "Monge"
  }
}
```

#### JWT payload 📇

The access token has the following claims :

- **`sub` :** the user `id` (as per the [JWT specs](https://tools.ietf.org/html/rfc7519#section-4.1.2)).
- **`email` :** the user `email`.

## Authorization flow 🛡️

JSON Server Auth provides generic guards as **route middlewares**.

To handle common use cases, JSON Server Auth draws inspiration from **Unix filesystem permissions**, especialy the [numeric notation](https://en.wikipedia.org/wiki/File_system_permissions#Numeric_notation).

- We add **`4`** for **read** permission.
- We add **`2`** for **write** permission.

_Of course CRUD is not a filesystem, so we don't add 1 for execute permission._

Similarly to Unix, we then have three digits to match each user type :

- First digit are the permissions for the **resource owner**.
- Second digit are the permissions for the **logged-in users**.
- Third digit are the permissions for the **public users**.

For example, **`640`** means that only the owner can write the resource, logged-in users can read the resource, and public users cannot access the resource at all.

#### The resource owner 🛀

A user is the owner of a resource if that resource has a **`userId`** property that matches his `id` property. Example:

```js
// The owner of
{ id: 8, text: 'blabla', userId: 1 }
// is
{ id: 1, email: 'olivier@mail.com' }
```

Private guarded routes will use the JWT `sub` claim (which equals the user `id`) to check if the user actually owns the requested resource, by comparing `sub` with the `userId` property.

_Except for the actual `users` collection, where the JWT `sub` claim must match the `id` property._

### Guarded routes 🚥

Guarded routes exist at the root and can restrict access to any resource you put after them :

|    Route     | Resource permissions                                                                                 |
| :----------: | :--------------------------------------------------------------------------------------------------- |
| **`/664/*`** | User must be logged to _write_ the resource. <br> Everyone can _read_ the resource.                  |
| **`/660/*`** | User must be logged to _write_ or _read_ the resource.                                               |
| **`/644/*`** | User must own the resource to _write_ the resource. <br> Everyone can _read_ the resource.           |
| **`/640/*`** | User must own the resource to _write_ the resource. <br> User must be logged to _read_ the resource. |
| **`/600/*`** | User must own the resource to _write_ or _read_ the resource.                                        |
| **`/444/*`** | No one can _write_ the resource. <br> Everyone can _read_ the resource.                              |
| **`/440/*`** | No one can _write_ the resource. <br> User must be logged to _read_ the resource.                    |
| **`/400/*`** | No one can _write_ the resource. <br> User must own the resource to _read_ the resource.             |

#### Examples

- Public user (not logged-in) does the following requests : 

| _Request_                               | _Response_         |
| :-------------------------------------- | :----------------- |
| `GET /664/posts`                        | `200 OK`           |
| `POST /664/posts`<br>`{text: 'blabla'}` | `401 UNAUTHORIZED` |

- Logged-in user with `id: 1` does the following requests :

| _Request_                                                  | _Response_      |
| :--------------------------------------------------------- | :-------------- |
| `GET /600/users/1`<br>`Authorization: Bearer xxx.xxx.xxx`  | `200 OK`        |
| `GET /600/users/23`<br>`Authorization: Bearer xxx.xxx.xxx` | `403 FORBIDDEN` |

### Setup permissions 💡

Of course, you don't want to directly use guarded routes in your requests.
We can take advantage of [JSON Server custom routes feature](https://github.com/typicode/json-server#add-custom-routes) to setup resource permissions ahead.

Create a `routes.json` file :

```json
{
  "/users*": "/600/users$1",
  "/messages*": "/640/messages$1"
}
```

Then :

```bash
json-server db.json -m ./node_modules/json-server-auth -r routes.json
# with json-server installed globally and json-server-auth installed locally
```

##### 📢 but wait !

As a convenience, **`json-server-auth`** CLI allows you to define permissions in a more succinct way :

```json
{
  "users": 600,
  "messages": 640
}
```

Then :

```bash
json-server-auth db.json -r routes.json
# with json-server-auth installed globally
```

You can still add any other _normal_ custom routes :

```json
{
  "users": 600,
  "messages": 640,
  "/posts/:category": "/posts?category=:category"
}
```

## Module usage 🔩

If you go the programmatic way and [use JSON Server as a module](https://github.com/typicode/json-server#module), there is an extra step to properly integrate JSON Server Auth :

⚠️ You must bind the router property `db` to the created app, like the [JSON Server CLI does](https://github.com/typicode/json-server/blob/master/src/cli/run.js#L74), and you must apply the middlewares in a specific order.

```js
const jsonServer = require('json-server')
const auth = require('json-server-auth')

const app = jsonServer.create()
const router = jsonServer.router('db.json')

// /!\ Bind the router db to the app
app.db = router.db

// You must apply the auth middleware before the router
app.use(auth)
app.use(router)
app.listen(3000)
```

#### Permisssions Rewriter

The custom rewriter is accessible via a subproperty :

```js
const auth = require('json-server-auth')

const rules = auth.rewriter({
  // Permission rules
  users: 600,
  messages: 640,
  // Other rules
  '/posts/:category': '/posts?category=:category',
})

// You must apply the middlewares in the following order
app.use(rules)
app.use(auth)
app.use(router)
```

## TODO 📜

- [ ] Use JSON Server `id` and `foreignKeySuffix` parameters
- [ ] Handle query params in list requests to secure guarded routes more precisely
- [ ] Allow configuration of :
  - [ ] Users collection name
  - [ ] Minimum password length
  - [ ] JWT expiry time
  - [ ] JWT property name in response
- [ ] Implement JWT Refresh Token
- [ ] Possibility to disable password encryption ?


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [2.1.0](https://github.com/jeremyben/json-server-auth/compare/v2.0.2...v2.1.0) (2021-07-21)


### Features

* register and login return user data besides access token ([5af16c9](https://github.com/jeremyben/json-server-auth/commit/5af16c940e8a41b0bd81c478813827561eb2d5b9))


### Bug Fixes

* bin path and peer dependency version ([89af19e](https://github.com/jeremyben/json-server-auth/commit/89af19ef636136d8db389879857282cfc6a1636f))
* put and patch on user collection ([4382ef1](https://github.com/jeremyben/json-server-auth/commit/4382ef1a41dfa90734719eb2cc163c355ec0d733))

### [2.0.2](https://github.com/jeremyben/json-server-auth/compare/v2.0.1...v2.0.2) (2020-06-01)


### Bug Fixes

* allow other HTTP methods like OPTIONS to pass through guards ([7b8d1a0](https://github.com/jeremyben/json-server-auth/commit/7b8d1a0fe9d12b4d527b3a795d4aed9fdcf07961))

### [2.0.1](https://github.com/jeremyben/json-server-auth/compare/v2.0.0...v2.0.1) (2020-06-01)


### Bug Fixes

* allow user creation on users route after setting guards ([a29ec45](https://github.com/jeremyben/json-server-auth/commit/a29ec452141f79fc5967538d4a852b3462f2b928))

## [2.0.0](https://github.com/jeremyben/json-server-auth/compare/v1.2.1...v2.0.0) (2020-05-31)


### ⚠ BREAKING CHANGES

* well it's no longer a direct dependency

### Bug Fixes

* LF end of line ([ed2483c](https://github.com/jeremyben/json-server-auth/commit/ed2483c53e6082f5beed6c758f9080218643ecbd))
* move json-server as a peer depency, remove express as direct dependency (already in json-server) ([a4edcdc](https://github.com/jeremyben/json-server-auth/commit/a4edcdcdffcbb2015a43f5cb8c80190a112e5a41))

### [1.2.1](https://github.com/jeremyben/json-server-auth/compare/v1.2.0...v1.2.1) (2019-06-15)


### Bug Fixes

* unique constraint on user email ([eb47b25](https://github.com/jeremyben/json-server-auth/commit/eb47b252612628b03821876db62bf6ed5ba1490f))

### [1.2.0](https://github.com/jeremyben/json-server-auth/compare/v1.1.0...v1.2.0) (2019-05-11)


### Features

* custom rewriter is now accessible from the module root ([4fbdf70](https://github.com/jeremyben/json-server-auth/commit/4fbdf70bc72119ff79ee8a686162e62020a64bb8))

### Bug Fixes

* compiler errors and missing flag in cli bin ([51546eb](https://github.com/jeremyben/json-server-auth/commit/51546ebe05f1d1300b6debc7eb5e0850f4fd1add))
* users put route would not validate properly email and password ([d3b73d3](https://github.com/jeremyben/json-server-auth/commit/d3b73d3f6e9d5479de6ba6a07f0eabf17f0c2cf8))

### [1.1.0](https://github.com/jeremyben/json-server-auth/compare/v1.0.0...v1.1.0) (2018-12-24)


### Features

* convenient guards rewriter ([d41c29e](https://github.com/jeremyben/json-server-auth/commit/d41c29e6fb49a51df278f4ecfbe268c94596955b))

## 1.0.0 (2018-12-17)


### Features

* users router, guards router, cli ([deb84ab](https://github.com/jeremyben/json-server-auth/commit/deb84abd65fcc95c10c6b3e5968d9495e4acf0d4))


src/guards.ts
import { Router, Handler } from 'express'
import * as jwt from 'jsonwebtoken'
import * as jsonServer from 'json-server'
import { stringify, ParsedUrlQueryInput } from 'querystring'
import { JWT_SECRET_KEY } from './constants'
import { bodyParsingHandler, errorHandler, goNext } from './shared-middlewares'

/**
 * Logged Guard.
 * Check JWT.
 */
const loggedOnly: Handler = (req, res, next) => {
	if (
		req.method !== 'GET' &&
		req.method !== 'POST' &&
		req.method !== 'PUT' &&
		req.method !== 'PATCH' &&
		req.method !== 'DELETE'
	) {
		// We let pass the other methods (HEAD, OPTIONS)
		// as they are not handled by json-server router,
		// but maybe by another user-defined middleware
		next()
		return
	}

	const { authorization } = req.headers

	if (!authorization) {
		res.status(401).jsonp('Missing authorization header')
		return
	}

	const [scheme, token] = authorization.split(' ')

	if (scheme !== 'Bearer') {
		res.status(401).jsonp('Incorrect authorization scheme')
		return
	}

	if (!token) {
		res.status(401).jsonp('Missing token')
		return
	}

	try {
		jwt.verify(token, JWT_SECRET_KEY)
		// Add claims to request
		req.claims = jwt.decode(token) as any
		next()
	} catch (err) {
		res.status(401).jsonp((err as jwt.JsonWebTokenError).message)
	}
}

/**
 * Owner Guard.
 * Checking userId reference in the request or the resource.
 * Inherits from logged guard.
 */
const privateOnly: Handler = (req, res, next) => {
	loggedOnly(req, res, () => {
		const { db } = req.app
		if (db == null) {
			throw Error('You must bind the router db to the app')
		}

		// TODO: handle query params instead of removing them
		const path = req.url.replace(`?${stringify(req.query as ParsedUrlQueryInput)}`, '')
		const [, mod, resource, id] = path.split('/')

		// Creation and replacement
		// check userId on the request body
		if (req.method === 'POST') {
			// TODO: use foreignKeySuffix instead of assuming the default "Id"
			const hasRightUserId = String(req.body.userId) === req.claims!.sub

			// No userId reference when creating a new user (duh)
			const isUserResource = resource === 'users'

			if (hasRightUserId || isUserResource) {
				next()
			} else {
				res.status(403).jsonp(
					'Private resource creation: request body must have a reference to the owner id'
				)
			}

			return
		}

		if (req.method === 'PUT') {
			// TODO: use foreignKeySuffix instead of assuming the default "Id"
			const hasRightUserId = String(req.body.userId) === req.claims!.sub

			const isUserResourceAndIsRightId = resource === 'users' && id === req.claims!.sub

			if (hasRightUserId || isUserResourceAndIsRightId) {
				next()
			} else {
				res.status(403).jsonp(
					'Private resource replacement: request body must have a reference to the owner id'
				)
			}

			return
		}

		// Query and update
		// check userId on the resource
		if (req.method === 'GET' || req.method === 'PATCH' || req.method === 'DELETE') {
			let hasRightUserId: boolean

			// TODO: use foreignKeySuffix instead of assuming the default "Id"
			if (id) {
				const entity: Record<string, any> = db.get(resource).getById(id).value()
				// get id if we are in the users collection
				const userId = resource === 'users' ? entity.id : entity.userId

				hasRightUserId = String(userId) === req.claims!.sub
			} else {
				const entities: Record<string, any>[] = db.get(resource).value()

				// TODO: Array.every() for properly secured access.
				// Array.some() is too relax, but maybe useful for prototyping usecase.
				// But first we must handle the query params.
				hasRightUserId = entities.some(
					(entity) => String(entity.userId) === req.claims!.sub
				)
			}

			if (hasRightUserId) {
				next()
			} else {
				res.status(403).jsonp(
					'Private resource access: entity must have a reference to the owner id'
				)
			}

			return
		}

		// We let pass the other methods (HEAD, OPTIONS)
		// as they are not handled by json-server router,
		// but maybe by another user-defined middleware
		next()
	})
}
// tslint:enable

/**
 * Forbid all methods except GET.
 */
const readOnly: Handler = (req, res, next) => {
	if (req.method === 'GET') {
		next()
	} else {
		res.status(403).jsonp('Read only')
	}
}

type ReadWriteBranch = ({ read, write }: { read: Handler; write: Handler }) => Handler

/**
 * Allow applying a different middleware for GET request (read) and others (write)
 * (middleware returning a middleware)
 */
const branch: ReadWriteBranch =
	({ read, write }) =>
	(req, res, next) => {
		if (req.method === 'GET') {
			read(req, res, next)
		} else {
			write(req, res, next)
		}
	}

/**
 * Remove guard mod from baseUrl, so lowdb can handle the resource.
 */
const flattenUrl: Handler = (req, res, next) => {
	// req.url is writable and used for redirection,
	// but app.use() already trim baseUrl from req.url,
	// so we use app.all() that leaves the baseUrl with req.url,
	// so we can rewrite it.
	// https://stackoverflow.com/questions/14125997/

	req.url = req.url.replace(/\/[640]{3}/, '')
	next()
}

/**
 * Guards router
 */
export default Router()
	.use(bodyParsingHandler)
	.all('/666/*', flattenUrl)
	.all('/664/*', branch({ read: goNext, write: loggedOnly }), flattenUrl)
	.all('/660/*', loggedOnly, flattenUrl)
	.all('/644/*', branch({ read: goNext, write: privateOnly }), flattenUrl)
	.all('/640/*', branch({ read: loggedOnly, write: privateOnly }), flattenUrl)
	.all('/600/*', privateOnly, flattenUrl)
	.all('/444/*', readOnly, flattenUrl)
	.all('/440/*', loggedOnly, readOnly, flattenUrl)
	.all('/400/*', privateOnly, readOnly, flattenUrl)
	.use(errorHandler)

/**
 * Transform resource-guard mapping to proper rewrite rule supported by express-urlrewrite.
 * Return other rewrite rules as is, so we can use both types in routes.json.
 * @example
 * { 'users': 600 } => { '/users*': '/600/users$1' }
 */
export function parseGuardsRules(resourceGuardMap: { [resource: string]: any }) {
	return Object.entries(resourceGuardMap).reduce((routes, [resource, guard]) => {
		const isGuard = /^[640]{3}$/m.test(String(guard))

		if (isGuard) {
			routes[`/${resource}*`] = `/${guard}/${resource}$1`
		} else {
			// Return as is if not a guard
			routes[resource] = guard
		}

		return routes
	}, {} as Parameters<typeof jsonServer.rewriter>[0])
}

/**
 * Conveniant method to use directly resource-guard mapping
 * with JSON Server rewriter (which itself uses express-urlrewrite).
 * Works with normal rewrite rules as well.
 */
export function rewriter(resourceGuardMap: { [resource: string]: number }): Router {
	const routes = parseGuardsRules(resourceGuardMap)
	return jsonServer.rewriter(routes)
}
